### DISCUSSION TITLE
Possible use of Briefcase inside other projects?

### DISCUSSION THREAD

**Discussion by konstantinkopka** [beeware/briefcase#928] [General]
Hi @freakboy3742 
I really love your vision for BeeWare and 100% agree that it should be way more straightforward to build Python apps for Mobile and Web. 

I love the [Godot Game Engine](https://godotengine.org/) and would really like to use some Python with it instead of the built-in GDScript language (which is fine but lacks a thriving ecosystem of libraries).

@touilleMan is [attempting to support this](https://github.com/touilleMan/godot-python)  and it looks like exporting to Desktop platforms is possible but Mobile and Web seem like a very daunting task.

So I was wondering: Is there any potential for embedding/using what's possible with Briefcase together with godot-python?

Maybe there's some potential for cross-collaboration here as both projects aim to solve a very similar issue: Bringing Python to more platforms and achieving new things with it.

**Comment by freakboy3742** [comment]
I don't know enough about Godot (or Godot-python) to be able to comment about what is or isn't possible. As a broad goal, I'd say that sounds a use case that would be *desirable* for Briefcase to support - and if it can't, then I'm interested in exploring what pieces are missing that are the cause of problems.

The one thing to keep in mind, though, is that Briefcase won't magically add mobile (or web) support to a GUI toolkit that doesn't already support mobile (or web). It simplifies the process of getting Python code onto a mobile platform - but it won't make binary code automatically mobile compatible. The limitations here are entirely up to Godot. If Godot doesn't already provide iOS support, Briefcase won't make Godot iOS compatible.

However, assuming that's not a problem: As it currently stands, Briefcase is essentially a templating engine. The difference between HelloWorld.app and any other app is the list of dependencies and the app's code itself. The core of a packaged app bundle is shared between every possible app.

If a Godot app can be made to fit into this exiting framework (i.e., you can "pip install Godot" and then have some user-space code that runs `import Godot` or similar), then it is entirely possible that Godot will fit into the framework that Briefcase already provides.

If a Godot app needs it's own special infrastructure to start a Godot app, then it may be possible to support Godot through the use of a platform plugin. Essentially "Godot" becomes a target packaging format for an app. 

If someone is interested in experimenting here, he next step is to develop a "reference" app - an example application on one platform. If you are able to build this reference app using a "briefcase new" project with "no GUI" as a starting point, then it should be possible to modify the "new project" template to provide the scaffold for a Godot project.

However, if you need to modify the basic structure of the new Briefcase project, then we're likely looking at needing a custom platform plugin.

Either way, I'm happy to provide guidance and feedback from the Briefcase side of the equation.

**Reply by konstantinkopka** [reply]
@freakboy3742 Thanks, that sounds very promising :)
Unfortunately I also know too little about how Godot and godot-python work.
@touilleMan is probably the most knowledgeable person on the planet to talk about this, since he not only builds godot-python but is also a core contributor to the Godot project itself and knows its ins and outs.

> The limitations here are entirely up to Godot. If Godot doesn't already provide iOS support, Briefcase won't make Godot iOS compatible.

I think this shouldn't pose a problem. Godot itself already provides out of the box support for exporting to Desktop, Mobile, Web and even Game Consoles (with 3rd party help), provided you use the built-in GDScript language. Godot also has a really amazing GUI framework and visual editor to work in, which is one of the big reasons why I want to use it (imho it blows other UI toolkits like Qt or WxWidgets out of the water)

If I understood touilleMan correctly, godot-python will use the new GDExtension system and other things™ such as compiling to C directly with Cython, but I don't understand the details of it, it seemed pretty advanced.

Unfortunately I can't provide much technical contribution, since I am way out of my depth with this (I am just a game developer with years of mostly C# experience). I can, however, say, that it would be absolutely awesome to one day be able to use Godot with Python and export apps to all major platforms, including Web & Mobile. It would be a dream come true :)
